### ISSUE TITLE
PME for Magnetostatics

### ISSUE THREAD

**Issue by PicoCentauri** [lab-cosmo/torch-pme#54]
Besides, computing the potential for point charges another very important interaction is the one between point dipoles according to

$$
U^{Dip}(\vec{r}) = D \cdot \left( \frac{(\vec{\mu}_i \cdot \vec{\mu}_j)}{r^3} - \frac{3  (\vec{\mu}_i \cdot \vec{r})  (\vec{\mu}_j \cdot \vec{r}) }{r^5} \right)
$$

where $`r=|\vec{r}|$ and the prefactor $D$ is given by $D =μ_0/(4π)$. Here, $μ_0$ is the vacuum permittivity.

Analogous to a PME algorithm for point charges there also exist fast solvers for dipolar interaction usually called dipolar PME algorithms as described for example in [Cérda 2008 ](https://pubs.aip.org/aip/jcp/article-abstract/129/23/234104/349096/P3M-algorithm-for-dipolar-interactions?redirectedFrom=fulltext).

@sirmarcel and I saw systems where it seemed crucial to also add point dipoles on top of point charges to model the energy and forces of such structures accurately. Therefore, such calculators could be a very powerful and logical extension of `torch-pme`.

**Comment by chiang-yuan** [comment]
Thanks for developing and open-sourcing this. I would be interested in helping to make this happen earlier and have many immediate needs for different systems. Is anyone actively working on this? Happy to jump into it to lower the barrier to be included as soon as possible. :)

**Comment by ceriottm** [comment]
hello! great to see involvement already this early after we release the first beta.
I would say that you could start helping merge #128 so you can also ask questions about the code structure and philosophy. There will be a few more adjustments over the next weeks as we have already a TODO list before the accompanying paper gets published, but I'd say that if in the meantime you want to fork, create a magnetostatics branch and open a draft PR, we'll be responsive to questions ^_^.

**Comment by PicoCentauri** [comment]
You can try starting, which is also a good test how well described the code is. If you need some kickstart, we can also have an independent chat. Email me in case.

**Comment by chiang-yuan** [comment]
@ceriottm @PicoCentauri Yes I have used your code calculate the electrostatics and found it is very reliable and smooth! Multipole expansion is the natural extension and much needed. 

The code base design is a bit intricate but the I found document is great and code is self-explanatory. Maybe a high level question is --- Is appending multipoles into a single pass or separating them into multiple class more preferred?

**Comment by PicoCentauri** [comment]
Yes the code still has some not super clear parts. See also #57.

I would go for a different class called `DipolarP3M`, which should maybe look like

```python
import torch
from torchpme.potentials import Potential


class DipolarP3M:
    def __init__(
        self,
        potential: Potential,
        mesh_spacing: float,
        interpolation_nodes: int = 4,
        full_neighbor_list: bool = False,
        prefactor: float = 1.0,
    ): ...

    def forward(
        self,
        dipoles: torch.Tensor,
        cell: torch.Tensor,
        positions: torch.Tensor,
        neighbor_indices: torch.Tensor,
        neighbor_distances: torch.Tensor,
    ): ...
```

If we want to implement more then just P3M for dipoles it may makes sense to write another base class and put everything in a subfolder `magnetostatics`. But we will see how this evolves.

**Comment by E-Rum** [comment]
I want to start implementing the core classes for dipole-dipole interactions, beginning with a direct “potential” since it is the easiest to implement.

The question is: what do you think is the optimal way to store the building blocks? For charge interactions, we can introduce the notion of a scalar potential and calculate the energy as the product of this potential with a scalar charge. For dipoles, the final energy is obtained via a dot product.

Thus, we can either compute the energy directly or, if we aim to preserve the concept of "generalized potential" * "generalized charge" (with the dipole moment acting as the generalized charge in this case), the potential would need to be represented as a vector. However, I am unsure whether this approach is reasonable.

What do you think?

**Comment by sirmarcel** [comment]
I think for ML purposes it is preferable to have access to the per-atom vector potential, in the same way that we offer access to the per-atom scalar potential.
